### PR TITLE
Add phone exists page for register journey

### PR DESCRIPTION
**Pull request by wilpow-github**
### Context

As an overseas qualified teacher
I want to sign in/register a Teaching services account
So that I can apply for QTS

### Changes proposed in this pull request

Modify the /sign-in/register/phone-confirmation page to check if there is an existing User after verifying the PIN code
Redirect to /sign-in/register/phone-exists (a new page) that matches the designs in the prototype.

### Checklist

-   [x] Attach to Trello card
-   [x] Rebased master
-   [x] Cleaned commit history
-   [x] Tested by running locally
-   [x] Reminder created to manually clean any removed app settings post deployment

### CODE REVIEW

File: dotnet-authserver/src/TeacherIdentity.AuthServer/AuthenticationState.cs [hunk 1]
```diff
@@ -678,6 +670,26 @@ private bool ShouldRedirectToLandingPage()
 
         return !OAuthState.HasAnyScope(ignoredScopes) && !UserId.HasValue;
     }
+
+    private void UpdateAuthenticationStateWithUserDetails(User user)
+    {
+        UserId = user.UserId;
+        EmailAddress = user.EmailAddress;
+        MobileNumber = user.MobileNumber;
```
Comment (wilpow-github) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/AuthenticationState.cs:678]: Added updates to MobileNumber and EmailAddress here. There's a debug assertion in OnEmailAddressVerified that indicates we may have different cases, but assuming that's ok.
`Debug.Assert(user.EmailAddress.Equals(EmailAddress, StringComparison.OrdinalIgnoreCase));`

File: dotnet-authserver/src/TeacherIdentity.AuthServer/AuthenticationState.cs [hunk 2]
```diff
@@ -335,6 +322,11 @@ public void OnMobileNumberVerified(User? user = null)
         }
 
         MobileNumberVerified = true;
+
+        if (user is not null)
+        {
+            UpdateAuthenticationStateWithUserDetails(user);
```
Comment (wilpow-github) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/AuthenticationState.cs:328]: If email address entered in previous step is different to existing user address do we need to redirect to register/choose-email page?

Comment (wilpow-github) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/AuthenticationState.cs:328]: Also assuming we want to do similar sign-in for register/email-confirmation

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/PhoneConfirmation.cshtml.cs [hunk 3]
```diff
@@ -43,7 +49,25 @@ public async Task<IActionResult> OnPost()
             return await HandlePinVerificationFailed(pinVerificationFailedReasons);
         }
 
-        HttpContext.GetAuthenticationState().OnMobileNumberVerified();
+        var authenticationState = HttpContext.GetAuthenticationState();
+        var permittedUserTypes = authenticationState.GetPermittedUserTypes();
+
+        var user = await _dbContext.Users
+            .Where(u => (u.MobileNumber ?? "").EndsWith(NormaliseMobileNumber(MobileNumber!)))
```
Comment (wilpow-github) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/PhoneConfirmation.cshtml.cs:56]: Unfortunately we have to do some formatting of the mobile number before storing or this lookup will be inefficient. At the moment formatting should be done when the mobile number is saved to the user model (not yet implemented), but can perhaps be done earlier in e.g. register/phone

File: dotnet-authserver/tests/TeacherIdentity.AuthServer.TestCommon/TestData.CreateUser.cs [hunk 4]
```diff
@@ -45,10 +46,13 @@ public partial class TestData
                 throw new ArgumentException($"{userType} users should not have {nameof(User.CompletedTrnLookup)} set.");
             }
 
+            var normalisedMobileNumber = Regex.Replace(Faker.Phone.Number(), @"(?<!^)[^\d]", "");
```
Comment (wilpow-github) [inline dotnet-authserver/tests/TeacherIdentity.AuthServer.TestCommon/TestData.CreateUser.cs:49]: Store the number as digits only, except for possible leading +

File: dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/EndpointTests/SignIn/Register/PhoneConfirmationTests.cs [hunk 5]
```diff
@@ -298,4 +299,112 @@ public async Task Post_BlockedClient_ReturnsTooManyRequestsStatusCode()
...
+        // Assert
+        Assert.Equal(StatusCodes.Status302Found, (int)response.StatusCode);
+        Assert.StartsWith("/sign-in/register/phone-exists", response.Headers.Location?.OriginalString);
+
+        Assert.True(authStateHelper.AuthenticationState.MobileNumberVerified);
+        Assert.NotNull(authStateHelper.AuthenticationState.UserId);
+    }
+
+    [Fact]
+    public async Task Post_ValidPinForAdminScopeWithAdminUser_UpdatesAuthenticationStateSignsInAndRedirects()
```
Comment (wilpow-github) [inline dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/EndpointTests/SignIn/Register/PhoneConfirmationTests.cs:333]: Should we even allow requests with admin scopes to enter register journey?

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/PhoneConfirmation.cshtml.cs [hunk 6]
```diff
@@ -57,4 +81,14 @@ public override void OnPageHandlerExecuting(PageHandlerExecutingContext context)
...
     }
+
+    private string NormaliseMobileNumber(string mobileNumber)
+    {
+        return new string(RemoveUkCountryCode(mobileNumber).Where(char.IsDigit).ToArray()).TrimStart('0');
+    }
+
+    private string RemoveUkCountryCode(string mobileNumber)
+    {
+        return Regex.Replace(mobileNumber, @"^(\+44|0044)", "");
```
Comment (wilpow-github) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/PhoneConfirmation.cshtml.cs:92]: For national numbers user can enter either 07877 722 722 or 00447877 722 722 and we want to treat as equal. Not the case for international numbers

File: dotnet-authserver/tests/TeacherIdentity.AuthServer.TestCommon/TestData.CreateUser.cs [hunk 7]
```diff
@@ -45,10 +46,13 @@ public partial class TestData
                 throw new ArgumentException($"{userType} users should not have {nameof(User.CompletedTrnLookup)} set.");
             }
 
+            var normalisedMobileNumber = Regex.Replace(Faker.Phone.Number(), @"^[^\d\+]|(?<=.)[^\d]", "");
```
Comment (wilpow-github) [inline dotnet-authserver/tests/TeacherIdentity.AuthServer.TestCommon/TestData.CreateUser.cs:49]: Store the number as digits only, except for possible leading +

File: dotnet-authserver/src/TeacherIdentity.AuthServer/AuthenticationState.cs (hunk 2 above)
Comment (gunndabad) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/AuthenticationState.cs:328]: There are/will be separate 'edge case' tickets for dealing with both of those.

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/PhoneConfirmation.cshtml.cs (hunk 3 above)
Comment (gunndabad) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/PhoneConfirmation.cshtml.cs:56]: I think as you have it here is best; ideally the value in `AuthenticationState` is as the user entered it, at least for the registration process.

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/PhoneConfirmation.cshtml.cs [hunk 8]
```diff
@@ -57,4 +81,14 @@ public override void OnPageHandlerExecuting(PageHandlerExecutingContext context)
             context.Result = new RedirectResult(_linkGenerator.RegisterPhone());
         }
     }
+
+    private string NormaliseMobileNumber(string mobileNumber)
+    {
+        return new string(RemoveUkCountryCode(mobileNumber).Where(char.IsDigit).ToArray()).TrimStart('0');
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/PhoneConfirmation.cshtml.cs:87]:
```suggestion
        return new string(RemoveUkCountryCode(mobileNumber).Where(char.IsAsciiDigit).ToArray()).TrimStart('0');
```

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Models/Mappings/UserMapping.cs [hunk 9]
```diff
@@ -27,6 +27,7 @@ public void Configure(EntityTypeBuilder<User> builder)
         builder.Property(u => u.IsDeleted).IsRequired().HasDefaultValue(false);
         builder.Property(u => u.TrnLookupStatus);
         builder.Property(u => u.MergedWithUserId);
+        builder.Property(u => u.MobileNumber).HasMaxLength(100);
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Models/Mappings/UserMapping.cs:30]: I think we need an unique index on `MobileNumber` too (with an `is_deleted = false` condition)

File: dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/EndpointTests/SignIn/Register/PhoneConfirmationTests.cs (hunk 5 above)
Comment (gunndabad) [reply dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/EndpointTests/SignIn/Register/PhoneConfirmationTests.cs:333]: We'll need a set of filters for `/Trn` and `/Register` eventually that ensure the initial request is 'compatible'. We're unlikely to support registering admin users so I don't think we need this scenario.

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Models/Mappings/UserMapping.cs [hunk 10]
```diff
@@ -27,6 +27,8 @@ public void Configure(EntityTypeBuilder<User> builder)
...
+        builder.HasIndex(u => u.MobileNumber).IsUnique();
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Models/Mappings/UserMapping.cs:31]:
```suggestion
        builder.HasIndex(u => u.MobileNumber).IsUnique().HasFilter("is_deleted = false and mobile_number is not null");
```

**Review by gunndabad — APPROVED** [review]

**Pull request merged**